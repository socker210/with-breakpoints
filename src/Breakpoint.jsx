import React from 'react'
import PropTypes from 'prop-types'
import MatchMediaquery from 'matchmediaquery'
import json2mq from 'json2mq'

class Breakpoint extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      mm: undefined,
      matches: props.defaultMatches
    }

    this.updateMatches = this.updateMatches.bind(this)
  }

  componentWillMount () {
    if (typeof window !== 'object' || typeof window.matchMedia !== 'function') return

    const { mq } = this.props
    const mm = MatchMediaquery(typeof mq === 'string' ? mq : json2mq(mq))

    mm.addListener(this.updateMatches)

    this.setState({
      mm: mm,
      matches: mm.matches
    })
  }

  componentWillUnmount () {
    this._unmount = true

    this.state.mm.removeListener(this.updateMatches)
  }

  updateMatches ({ matches }) {
    if (this._unmount) return

    this.setState({ matches }, () => {
      if (typeof this.props.onChange === 'function') this.props.onChange(matches)
    })
  }

  render () {
    const { matches } = this.state
    const { children } = this.props

    if (typeof children === 'function') return children(matches)

    return matches ? children || null : null
  }
}

Breakpoint.defaultProps = {
  mq: 'all',
  defaultMatches: true
}

Breakpoint.propTypes = {
  mq: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array
  ]),
  onChange: PropTypes.func,
  defaultMatches: PropTypes.bool
}

export default Breakpoint