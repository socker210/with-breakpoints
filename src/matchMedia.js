class MatchMedia {
  constructor (mq) {
    this.mm = typeof window === 'undefined' || typeof window.matchMedia === 'undefined' ? undefined : window.matchMedia(mq)
  }

  addListener (func) {
    if (this.mm) {
      this.func = func

      this.mm.addListener(func)
    }
  }

  removeListener () {
    if (this.mm && this.func) this.mm.removeListener(this.func)
  }
}

export default function matchMedia (mq) {
  return new MatchMedia(mq)
}