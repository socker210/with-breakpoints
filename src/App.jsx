import React from 'react'
import { render } from 'react-dom'
import Breakpoint, { withBreakpoint } from '.'

const ExampleOfBreakpoint = () => (
  <>
    <Breakpoint mq='screen and (min-width: 768px)'>
      <h3>Screen size bigger than 767</h3>
    </Breakpoint>
    <Breakpoint mq='screen and (max-width: 767px)'>
      <h3>Screen size smaller than 768</h3>
    </Breakpoint>
  </>
)

/* eslint-disable react/prop-types */
const ExampleOfWithBreakpoint = ({ pc, mobile }) => (
  <>
    { pc ? 'PC' : mobile ? 'Mobile' : null }
  </>
)

const mqs = [
  { name: 'pc', mq: 'screen and (min-width: 768px)' },
  { name: 'mobile', mq: 'screen and (max-width: 767px)' }
]
const HOC = withBreakpoint(mqs)(ExampleOfWithBreakpoint)

const App = () => (
  <>
    <h1>Breakpoint</h1>
    <ExampleOfBreakpoint />
    <br /><br /><br />
    <h1>withBreakpoint</h1>
    <HOC />
  </>
)

const container = document.getElementById('react-root')

render(<App />, container)