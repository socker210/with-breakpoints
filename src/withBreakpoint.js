import React from 'react'
import hoistNonReactStatics from 'hoist-non-react-statics'
import json2mq from 'json2mq'
import matchMedia from './matchMedia'

const requiredKeys = ['name']

function validateMq (mq) {
  if (typeof mq !== 'object') throw new Error('Invalid type of \'mqs\'\'s element: must be an object')

  const providedKeys = Object.keys(mq)
  const nonProvidedKeys = requiredKeys.filter(key => providedKeys.indexOf(key) === -1)

  return [
    !nonProvidedKeys.length,
    nonProvidedKeys
  ]
}

function createHOC (WrappedComponent, mqs) {
  class HOC extends React.Component {
    constructor (props) {
      super(props)

      const state = {}

      mqs.map(mq => {
        const name = mq.name
        const query = mq.mq ? (typeof mq.mq === 'string' ? mq.mq : json2mq(mq)) : 'all'
        const mm = matchMedia(query)
        const initialMatches = mm.getMatches(mq.defaultMatches)
        const listener = ({ matches }) => this.setState(name, matches)

        mm.addListener(listener)

        state[name] = { matches: initialMatches }
        state[name].removeListener = mm.removeListener(listener)
      })

      this.state = state
    }

    componentWillUnmount () {
      this._unmount = true

      Object.keys(this.state)
        .map(mq => mq.removeListener())
    }

    getProps () {
      const state = Object.keys(this.state)
        .reduce((container, name) => {
          container[name] = this.state[name].matches

          return container
        }, {})

      return {
        ...this.props,
        ...state
      }
    }

    render () {
      const props = this.getProps()

      return React.isValidElement(WrappedComponent) ? WrappedComponent : <WrappedComponent {...props} />
    }
  }

  return HOC
}

export default mqs => WrappedComponent => {
  if (!Array.isArray(mqs)) throw new Error('Invalid type of \'mqs\': must be an array')

  mqs.map(mq => {
    const [isSuccess, nonProvidedKeys] = validateMq(mq)

    if (!isSuccess) throw new Error(`Cannot read property '${nonProvidedKeys.join(', ')}'`)
  })

  const Enhance = createHOC(WrappedComponent, mqs)

  hoistNonReactStatics(Enhance, WrappedComponent)

  return Enhance
}