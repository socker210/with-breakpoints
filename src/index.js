import Breakpoint from './Breakpoint'
import withBreakpoint from './withBreakpoint'

export default Breakpoint
export {
  withBreakpoint
}