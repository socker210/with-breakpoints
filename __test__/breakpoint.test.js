import React from 'react'
import { shallow } from 'enzyme'
import { withBreakpoint } from '../src'

const Hello = () => (
  <h1>Hello</h1>
)

describe('withBreakpoint', () => {
  it('render', () => {
    const Component = <h1>HAHA</h1>
    const mqs = [
      { name: 'a', mq: 'screen and (max-width: 1200px)', defaultMatches: false }
    ]

    const HOC = withBreakpoint(mqs)(Hello)
    const wrapper = shallow(<HOC />)

    expect(wrapper).toMatchSnapshot()
  })
})