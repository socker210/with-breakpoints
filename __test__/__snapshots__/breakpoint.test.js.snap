// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withBreakpoint render 1`] = `
<Hello
  a={false}
/>
`;
